RDPQ basic changes

adamcate opened the pull request:
This is a straightforward transition to RDPQ with no TMEM optimizations, etc. The main improvements at the moment are the ability to use and automatically configure multiple texture formats-- including color indexed formats--using `format_handling.h`'s `format_set_render_mode`. I think this needs more testing before a merge, and tiled_cached still uses the old rendering system, hence why this is a draft PR

adamcate committed:
replace display_context_t with surface_t *

Commit:
Add Tiled RDP support for multiple formats

Commit:
Reset mode at the end of tiled_render_rdp

Commit:
Fix width calculation code

Commit:
Add multiple formats to animated_sprite_draw

Commit:
Add format_set_render_mode for
cleaner code

Commit:
change format_set_render_mode to support mirroring

Commit:
fix using hslices instead of vslices

Commit:
update sprite_batch for rdpq

Commit:
update documentation

stefanmielke commented:
Looks really good!

A few comments:
- Do we plan to expose `format_set_render_mode`? If so, we may want to consider adding to the README.
- Some of the variables with `_0` and `_1` _could_ be replaced with [position_int](https://github.com/stefanmielke/libdragon-extensions/blob/main/include/position_int.h) and `width`/`height` with [size](https://github.com/stefanmielke/libdragon-extensions/blob/main/include/size.h) (not necessary, more like a cleanup really, and for `size` we may want a `size_int`).
- We could add `deprecated` to `tiled_cached` if we feel comfortable removing it in a future version (or at least not giving any more time to improve it).
- Some of the formatting is weird (like mixed tabs), you could configure your editor to use the [.clang-format](https://github.com/stefanmielke/libdragon-extensions/blob/main/.clang-format) file we have on the root to fix that.

adamcate commented:
> Looks really good!
> 
> 
> 
> A few comments:
> 
> - Do we plan to expose `format_set_render_mode`? If so, we may want to consider adding to the README.
> 
> - Some of the variables with `_0` and `_1` _could_ be replaced with [position_int](https://github.com/stefanmielke/libdragon-extensions/blob/main/include/position_int.h) and `width`/`height` with [size](https://github.com/stefanmielke/libdragon-extensions/blob/main/include/size.h) (not necessary, more like a cleanup really, and for `size` we may want a `size_int`).
> 
> - We could add `deprecated` to `tiled_cached` if we feel comfortable removing it in a future version (or at least not giving any more time to improve it).
> 
> - Some of the formatting is weird (like mixed tabs), you could configure your editor to use the [.clang-format](https://github.com/stefanmielke/libdragon-extensions/blob/main/.clang-format) file we have on the root to fix that.

Thanks for the feedback. I initially added 'format_set_render_mode' for cleaner code but I think it would be great to expose that functionality to users. I'll test it and add it to the README. I'll definitely look into using positionint. I'll take a look at the formatting when I can as well.

Commit:
update formatting using .clang-format

Commit:
fix s_1 unused

Commit:
Add rect_int and size_int.

Commit:
Modify tiled_render_rdp to use size and position

Commit:
Change tile index type to int16_t

adamcate committed:
Update README to include format_handling.h

adamcate committed:
Fix README change

adamcate committed:
Misc. rewordings

I made some minor changes to the README for spelling and grammar to make things more clear for the reader.

Commit:
add scrolling capability

Commit:
Merge branch 'opengl' of github.com:pepsiman3/libdragon-extensions into opengl to reconcile changes

Commit:
mark tiled_cached as deprecated

Commit:
add csv_reader

Commit:
fix <stdint.h> not included in csv_reader

Commit:
fix mismatch of int and int16_t

Commit:
Fix screen variables uninitialized

Commit:
fix tex_coords uninitialized

Commit:
repair memory corruption bug in map array
due to only allocating memory for char

Commit:
fix csv_reader to treat index -1 as empty

Commit:
correct rightmost texture coordinates for loading

Commit:
Fix calculation of initial and final x and y

adamcate committed:
Update README

stefanmielke commented on include/tiled.h:
```diff
@@ -54,18 +69,21 @@ Tiled *tiled_init(MemZone *memory_pool, sprite_t *sprite, const char *map_path,
  * @param screen_rect
  *        Rect of the current screen. Used to cull tiles outside of the screen.
  */
```
I believe you forgot to add `view_position` to the doc here.

adamcate replied:
Good catch

stefanmielke commented on include/tiled_cached.h:
```diff
@@ -48,8 +48,9 @@ typedef struct {
  */
-TiledCached *tiled_cached_init(MemZone *memory_pool, sprite_t *sprite, const char *map_path,
-							   Size map_size, Size tile_size);
+__attribute__((deprecated("TiledCached is no longer actively supported"))) TiledCached *
```
I think it would be nice to say that's deprecated on the README as well.

adamcate replied:
Agreed, I'll work on it

stefanmielke commented on src/animated_sprite.c:
```diff
@@ -33,10 +34,23 @@ void animated_sprite_tick(AnimatedSprite *anim, float anim_rate) {
-						MIRROR_DISABLED);
+		surface_t spritesheet_surface = sprite_get_pixels(anim->sprite);
+
+		int tex_width = anim->sprite->width / anim->sprite->hslices;
```
Does it make sense to use `anim->size` here instead? Not sure if we would have a situation where the size of a slice would be different from its size.
Although, if it does, should we change the `is_intersecting` above to use this instead? Or maybe not even receive the size and just calculate it as here on `init`?

adamcate replied:
I was planning on completely overhauling the animation system in a separate PR to incorporate some of the testing I've been doing. I think all of those are good points, I'm just thinking that there might be situations where an animation might contain frames of different sizes or at least widths to save RDRAM usage

stefanmielke replied:
We can leave that this way then and come back later.

stefanmielke commented on src/sprite_batch.c:
```diff
@@ -14,16 +15,30 @@ SpriteBatch *sprite_batch_init(MemZone *memory_pool, sprite_t *sprite, size_t qt
+	format_set_render_mode(sprite_batch->sprite, false);
+
+	int tex_width = sprite_batch->sprite->width / sprite_batch->sprite->hslices;
+	int tex_height = sprite_batch->sprite->height / sprite_batch->sprite->vslices;
```
Same comment as on `AnimatedSprite`.

stefanmielke commented on src/tiled.c:
```diff
@@ -1,19 +1,55 @@
+					tiled->tile_size.width;                                                        \
+	int initial_y = (screen_rect.pos.y - tiled->offset.y - view_position.y) /                      \
+					tiled->tile_size.height;                                                       \
+	size_t final_x = ((screen_rect.pos.x + screen_rect.size.width - tiled->offset.x -              \
```
Since the `initial_x` and `initial_y` values can now be negative, should we do the same for `final_x` and `final_y`?

adamcate committed:
Update readme

Commit:
fix tiled documentation

Commit:
remove non-working formats for now

Commit:
Merge branch 'opengl' of github.com:pepsiman3/libdragon-extensions into opengl
